File: Greedy/TransformString/trans.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    int transform(string &s1, string &s2) {
        // code here
        int n1 = s1.size();
        int n2 = s2.size();
        if(n1 != n2)return -1;
        unordered_map<char,int>mp1,mp2;
        for(int i=0;i<n1;i++){
            mp1[s1[i]]++;
            mp2[s2[i]]++;
        }
        for(auto Pair: mp1){
            char ele = Pair.first;
            if(mp2.find(ele)==mp2.end())return -1;
            else if(mp1[ele]!=mp2[ele])return -1;
        }
        
        int i = n1 - 1;
        int j = n2 - 1;
        int ans = 0;
        
        while(i >= 0 && j >= 0) {
        
            if(s1[i] == s2[j]) {
                i--;
                j--;
            }
            else {
                i--;
                ans++;
            }
        }
        
        return ans;
    }
};

File: Greedy/TransformString/trans.py
class Solution:
    def transform(self, s1: str, s2: str) -> int:
        n1 = len(s1)
        n2 = len(s2)

        if n1 != n2:
            return -1

        mp1 = {}
        mp2 = {}

        for i in range(n1):
            mp1[s1[i]] = mp1.get(s1[i], 0) + 1
            mp2[s2[i]] = mp2.get(s2[i], 0) + 1

        for ele in mp1:
            if ele not in mp2:
                return -1

            if mp1[ele] != mp2[ele]:
                return -1

        i = n1 - 1
        j = n2 - 1
        ans = 0

        while i >= 0 and j >= 0:
            if s1[i] == s2[j]:
                i -= 1
                j -= 1
            else:
                i -= 1
                ans += 1

        return ans